### PR TITLE
Make getting and setting the RAND default method thread safe

### PR DESCRIPTION
**Pull request by levitte**
From a comment in #3136, where it was noticed that `RAND_set_rand_method()` and friends weren't thread safe.

### CODE REVIEW

File: crypto/rand/rand_lib.c [hunk 1]
```diff
@@ -12,27 +12,49 @@
...
 static ENGINE *funct_ref = NULL;
+static CRYPTO_RWLOCK *rand_engine_lock = NULL;
 #endif
 static const RAND_METHOD *default_RAND_meth = NULL;
+static CRYPTO_RWLOCK *rand_meth_lock = NULL;
+static CRYPTO_ONCE rand_lock_init = CRYPTO_ONCE_STATIC_INIT;
+
+DEFINE_RUN_ONCE_STATIC(do_rand_lock_init)
+{
+    OPENSSL_init_crypto(0, NULL);
```
Comment (mattcaswell) [inline crypto/rand/rand_lib.c]: Why is this necessary? It looks like a potential source of deadlocks

File: crypto/rand/rand_lib.c [hunk 2]
```diff
@@ -49,13 +71,18 @@ const RAND_METHOD *RAND_get_rand_method(void)
 #endif
             default_RAND_meth = RAND_OpenSSL();
     }
+    CRYPTO_THREAD_unlock(rand_meth_lock);
     return default_RAND_meth;
```
Comment (mattcaswell) [inline crypto/rand/rand_lib.c]: Are assignments atomic? This read of the variable is unprotected by a lock, so I wonder what happens if RAND_set_rand_method() is executed at the same time as this read is taking place.

**Review by mattcaswell — CHANGES_REQUESTED** [review]
I wonder what the performance impact of this is?

File: crypto/rand/rand_lib.c (hunk 2 above)
Comment (levitte) [inline crypto/rand/rand_lib.c]: It's protected by the read on line 58, no?

Comment (levitte) [inline crypto/rand/rand_lib.c]: Or oh, RAND_*set*_rand_method()!  Ok, I get what you mean

File: crypto/rand/rand_lib.c (hunk 1 above)
Comment (levitte) [inline crypto/rand/rand_lib.c]: Perhaps it isn't...  I copied this stoopidly from the one in `md_rand.c`...

**Comment by levitte** [comment]
Fixed all outstanding concerns

> I wonder what the performance impact of this is?

Good question

**Comment by mattcaswell** [comment]
Travis is showing a failure.

File: crypto/rand/rand_lib.c [hunk 3]
```diff
@@ -49,13 +72,19 @@ const RAND_METHOD *RAND_get_rand_method(void)
 #endif
             default_RAND_meth = RAND_OpenSSL();
     }
-    return default_RAND_meth;
+    tmp_meth = default_RAND_meth;
```
Comment (mattcaswell) [inline crypto/rand/rand_lib.c]: Does default_RAND_meth have to be declared volatile for this to work? Can the compiler "optimise" this solution away? I'm not really sure.

**Comment by levitte** [comment]
> Travis is showing a failure.

Yeah, I forgot to free the locks...

**Comment by levitte** [comment]
You know, this isn't done anywhere else (have a look at `crypto/rsa/rsa_lib.c` for example), and back in #3136, @richsalz makes a good point reminding us of #2244.  Closing this.

**Comment by mattcaswell** [comment]
I'm confused. The discussion in #2244 seems to suggest that this PR has it exactly right and is what we should be doing??

**Comment by levitte** [comment]
Maybe I'm confused too, I thought it about redoing the whole defaults stuff to happen mostly at compile time...

**Comment by mattcaswell** [comment]
Travis is now passing. LGTM. My only outstanding question is my one on volatile above.

File: crypto/rand/rand_lib.c (hunk 3 above)
Comment (kaduk) [inline crypto/rand/rand_lib.c]: I believe that the function call to unlock will provide a compiler-level barrier (and the lock itself will give a memory barrier at runtime).  Function calls are considered to have side effects, in the abstract machine.

**Review by kaduk — COMMENTED** [review]
> redoing the whole defaults stuff to happen mostly at compile time...

There's using compile-time initialization instead of CRYPTO_THREAD_run_once(), and then there's having the default_foo variable for what is the default method when a method is not explicitly specified.  The latter makes sense to allow changing at runtime.

**Comment by levitte** [comment]
So, errr, was that an approval hidden in there somewhere?

File: crypto/rand/rand_lib.c [hunk 4]
```diff
@@ -12,27 +12,50 @@
...
-
 #include <openssl/engine.h>
+#include "internal/thread_once.h"
 
 #ifndef OPENSSL_NO_ENGINE
 /* non-NULL if default_RAND_meth is ENGINE-provided */
 static ENGINE *funct_ref = NULL;
+static CRYPTO_RWLOCK *rand_engine_lock = NULL;
 #endif
 static const RAND_METHOD *default_RAND_meth = NULL;
```
Comment (kaduk) [inline crypto/rand/rand_lib.c]: In other projects I usually see a comment "requires rand_meth_lock for access" or semantically similar; do we want to do something like that?

**Review by kaduk — APPROVED** [review]
Sure, you can count it as an approval from me if you want.  (Not that it adds up to anything concrete, of course.)

**Review by mattcaswell — APPROVED** [review]

**Comment by levitte** [comment]
Merged.

master: 87975cfa91c3f06a52f2550ed2ef632644be6dde and 789a2b6250d5e05dfde6ce259e79ef8c172c9f3f
1.1.0: 67114d93b47aa02192272182595c2b5d1bba8298 and df13eb7ae70073de8c95ca3c7d2f587037582470

**Pull request closed without merging**